### PR TITLE
Ops can take several zero copy buffers (#4786)

### PR DESCRIPTION
**Pull request by SyrupThinker**
I'm unhappy about the use of `Box<[ZeroCopyBuf]>` instead of just a `&[ZeroCopyBuf]` (and so is clippy, 77 cases of boxed_local), because it can keep some `ZeroCopyBuf'`s alive longer than necessary.

The latter would require ZeroCopyBuf to be cloneable to extend its lifetime beyond the slice's for async operations, which requires low level changes to the memory management.
Alternatively this can probably be avoided by making it a `&mut [Option<ZeroCopyBuf>]` so the user can move the `ZeroCopyBuf` out of the slice, but this feels quite awkward.

Some feedback on how to proceed on this would be nice.
I doubt the three variants have very differing performance characteristics as they all introduce one heap allocation during `bindings::send`, and one free at varying times.

### CODE REVIEW

**Comment by SyrupThinker** [comment]
The following are the test results for NixOS 20.03, considering the CI test are currently cancelled due to the lint errors:
```
failures:
    benchmark_test
    upgrade_in_tmpdir

test result: FAILED. 176 passed; 2 failed; 15 ignored; 0 measured; 0 filtered out
```

- `benchmark_test` fails at parsing the strace output, still investigating that
- `upgrade_in_tmpdir` is a dynamic linker issue due to my OS and will probably succeed on other systems

**Comment by bartlomieju** [comment]
Thanks for the PR @SyrupThinker - it's touching a lot of core concepts and requires careful review. We're gonna revisit it after 1.0.0 is released.

**Comment by bartlomieju** [comment]
@SyrupThinker I added 1.1.0 milestone to this PR; can you resolve the conflicts and check benchmarks against `master` (`req/s` to make sure that there's no performance regression)?

**Comment by SyrupThinker** [comment]
Benchmarks before and after the change are very close to each other,
some have slightly better and some slightly worse results.
The variance is probably caused by noise.

<details>
<summary>Detailed benchmarks</summary>

### deno_tcp
#### pre
```
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   188.60us  131.53us  11.09ms   97.97%
    Req/Sec    24.25k     1.99k   27.41k    69.90%
Requests/sec:  48228.37
Transfer/sec:      2.35MB
```
#### post
```
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   177.11us   53.29us   3.15ms   88.33%
    Req/Sec    25.41k     0.98k   27.46k    80.60%
Requests/sec:  50548.04
Transfer/sec:      2.46MB
```
### http_benchmark testing DENO using net/http
#### pre
```
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   412.09us  150.04us  11.90ms   96.72%
    Req/Sec    11.43k   792.91    12.02k    93.53%
Requests/sec:  22741.09
Transfer/sec:      2.41MB
```
#### post
```
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   423.43us  164.85us  12.44ms   95.35%
    Req/Sec    11.14k     1.08k   12.20k    85.00%
Requests/sec:  22170.51
Transfer/sec:      2.35MB
```
-----
### http_proxy_benchmark testing DENO using net/tcp
#### pre
```
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   500.41us  106.47us   5.84ms   89.21%
    Req/Sec     9.64k   682.46    11.52k    72.64%
Requests/sec:  19185.66
Transfer/sec:      1.61MB
```
#### post
```
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   501.01us  114.09us   5.34ms   90.62%
    Req/Sec     9.64k   745.82    11.44k    70.50%
Requests/sec:  19184.69
Transfer/sec:      1.61MB
```
### http_benchmark testing deno_core_http_bench
#### pre
```
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   159.23us   44.50us   2.45ms   81.99%
    Req/Sec    27.78k     1.74k   30.40k    76.37%
Requests/sec:  55228.67
Transfer/sec:      2.69MB
```
#### post
```
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   158.78us   68.55us   7.07ms   95.04%
    Req/Sec    27.94k     1.49k   30.37k    72.25%
Requests/sec:  55552.10
Transfer/sec:      2.70MB
```
### http_benchmark testing NODE
#### pre
```
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   288.68us  329.19us  18.45ms   98.32%
    Req/Sec    16.91k     2.40k   20.02k    79.00%
Requests/sec:  33612.84
Transfer/sec:      3.56MB
```
#### post
```
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   290.45us  205.66us  11.95ms   95.34%
    Req/Sec    16.57k     2.03k   19.80k    84.58%
Requests/sec:  32938.02
Transfer/sec:      3.49MB
```
### http_proxy_benchmark testing NODE
#### pre
```
Requests/sec:   3523.67
Transfer/sec:    385.40KB
```
#### post
```
Requests/sec:   3615.45
Transfer/sec:    395.44KB
```
### http_proxy_benchmark testing NODE tcp
#### pre
```
Requests/sec:  21440.89
Transfer/sec:      1.80MB
```
#### post
```
Requests/sec:  21032.33
Transfer/sec:      1.77MB
```
### http_benchmark testing node_tcp.js
#### pre
```
Requests/sec:  59745.61
Transfer/sec:      2.91MB
```
#### post
```
Requests/sec:  58396.83
Transfer/sec:      2.84MB
```
### http_benchmark testing RUST hyper
#### pre
```
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   119.99us   46.23us   4.25ms   73.17%
    Req/Sec    30.40k     1.19k   35.85k    76.25%
Requests/sec:  60464.01
Transfer/sec:      5.07MB
```
#### post
```
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   122.35us   44.64us   2.42ms   68.21%
    Req/Sec    29.84k     1.31k   33.04k    75.06%
Requests/sec:  59204.15
Transfer/sec:      4.97MB
```

</details>

**Comment by bartlomieju** [comment]
@piscisaureus PTAL

**Comment by piscisaureus** [comment]
what is the purpose of using multiple ZeroCopy buffers?

**Comment by SyrupThinker** [comment]
As noted in #4786 there are cases where multiple distinct buffers are needed and copying buffers should be avoided for performance or security reasons.

**Comment by piscisaureus** [comment]
OK in that case: I find the extra heap allocation unfortunate. I'd suggest to pass &mut [ZeroCopyBuf]. The op could then take ownership of the buffers by using std::mem::take(). 

Of course an op that takes a variable number of buffers would still need to heap allocate, but that'd be better than penalizing all ops for the few that would need that.

**Comment by SyrupThinker** [comment]
> OK in that case: I find the extra heap allocation unfortunate. I'd suggest to pass &mut [ZeroCopyBuf]. The op could then take ownership of the buffers by using std::mem::take().

As far as I can see `ZeroCopyBuf` doesn't implement `Default`, and I also can't imagine a sensible default value for it, so I can't use `std::mem::take`. (Thats why I considered using an `Option<ZeroCopyBuf>` though.

**Comment by piscisaureus** [comment]
> As far as I can see ZeroCopyBuf doesn't implement Default, and I also can't imagine a sensible default value for it, so I can't use std::mem::take

That doesn't matter. You could "take" the entire slice. The slice implements default.

I'll write an example if this is unclear.

**Comment by SyrupThinker** [comment]
Using `std::mem::take` on a `&mut [ZeroCopyBuf]` causes it to fail because the slice itself doesn't implement `Default`.
```
the trait bound `[deno_core::zero_copy_buf::ZeroCopyBuf]: std::default::Default` is not satisfied
```
I think it'd need to be a `&mut &mut [ZeroCopySlice]` for that to work. I don't see how it would make sense with `&mut [ZeroCopySlice]`.

**Comment by piscisaureus** [comment]
> I think it'd need to be a `&mut &mut [ZeroCopySlice]` for that to work. I don't see how it would make sense with `&mut [ZeroCopySlice]`.

I think the problem is that `&mut [ZeroCopySlice]` is unsized, therefore it cannot be used with `take()`. `&mut &mut [ZeroCopySlice]` is possible but it'd be very impractical to use if you are actually dealing with a variable number of buffers -- it wouldn't be possible to create `Vec<ZeroCopyBuf>` and get a `&mut &mut ZeroCopyBuf` out of that.

So maybe we add some sort of wrapper type, e.g. `BufList`.

I wrote a quick prototype: 

```rust
use std::convert::From;
use std::convert::TryFrom;
use std::error::Error;
use std::fmt;
use std::iter::once;
use std::iter::IntoIterator;

pub struct ZeroCopyBuf {
    _private: (),
}

impl ZeroCopyBuf {
    pub fn mock() -> Self {
        Self { _private: () }
    }
}

pub enum BufList {
    None,
    One(ZeroCopyBuf),
    // Maybe add variant 'Two' if that ends up being used a lot.
    Any(Vec<ZeroCopyBuf>),
}

impl BufList {
    pub fn len(&self) -> usize {
        match self {
            BufList::None => 0,
            BufList::One(_) => 1,
            BufList::Any(l) => l.len(),
        }
    }
}

impl Default for BufList {
    fn default() -> Self {
        Self::None
    }
}

impl From<ZeroCopyBuf> for BufList {
    fn from(buf: ZeroCopyBuf) -> Self {
        Self::One(buf)
    }
}

impl<I: IntoIterator<Item = ZeroCopyBuf>> From<I> for BufList {
    fn from(iter: I) -> Self {
        let mut iter = iter.into_iter();
        if let Some(b0) = iter.next() {
            if let Some(b1) = iter.next() {
                Self::Any(once(b0).chain(once(b1)).chain(iter).collect())
            } else {
                Self::One(b0)
            }
        } else {
            Self::None
        }
    }
}

impl TryFrom<BufList> for () {
    type Error = BufCountError;
    fn try_from(list: BufList) -> Result<Self, Self::Error> {
        match list {
            BufList::None => Ok(()),
            BufList::Any(v) if v.len() == 0 => Ok(()),
            _ => Err(BufCountError::new(list, "0")),
        }
    }
}

impl TryFrom<BufList> for ZeroCopyBuf {
    type Error = BufCountError;
    fn try_from(list: BufList) -> Result<Self, Self::Error> {
        match list {
            BufList::One(b) => Ok(b),
            BufList::Any(v) if v.len() == 1 => Ok(v.into_iter().next().unwrap()),
            _ => Err(BufCountError::new(list, "1")),
        }
    }
}

impl TryFrom<BufList> for Option<ZeroCopyBuf> {
    type Error = BufCountError;
    fn try_from(list: BufList) -> Result<Self, Self::Error> {
        match list {
            BufList::None => Ok(None),
            BufList::One(b) => Ok(Some(b)),
            BufList::Any(v) if v.len() <= 1 => Ok(v.into_iter().next()),
            _ => Err(BufCountError::new(list, "0 or 1")),
        }
    }
}

impl From<BufList> for Vec<ZeroCopyBuf> {
    fn from(list: BufList) -> Self {
        match list {
            BufList::None => vec![],
            BufList::One(b) => vec![b],
            BufList::Any(v) => v,
        }
    }
}

pub struct BufCountError {
    list: BufList,
    expected_len: &'static str,
}
impl BufCountError {
    fn new(list: BufList, expected_len: &'static str) -> Self {
        Self { list, expected_len }
    }
}
impl fmt::Debug for BufCountError {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(formatter, "{}", self)
    }
}
impl fmt::Display for BufCountError {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            formatter,
            "Incorrect number of buffers: expected {}, but got {}.",
            self.list.len(),
            self.expected_len
        )
    }
}
impl Error for BufCountError {}

#[cfg(test)]
mod test {
    use super::*;
    use std::convert::TryInto;
    use std::iter::repeat_with;

    fn zero_buf_op(bufs: BufList) -> Result<(), BufCountError> {
        let _: () = bufs.try_into()?;
        Ok(())
    }

    fn one_buf_op(bufs: BufList) -> Result<(), BufCountError> {
        let _: ZeroCopyBuf = bufs.try_into()?;
        Ok(())
    }

    fn maybe_buf_op(bufs: BufList) -> Result<(), BufCountError> {
        let _: Option<ZeroCopyBuf> = bufs.try_into()?;
        Ok(())
    }

    fn multi_buf_op(bufs: BufList) {
        let _: Vec<ZeroCopyBuf> = bufs.into();
    }

    #[test]
    fn test_buf_list_ok() {
        let none_buf = Option::<ZeroCopyBuf>::None;
        let some_buf = Some(ZeroCopyBuf::mock());
        let one_buf = ZeroCopyBuf::mock();
        let multi_bufs = repeat_with(ZeroCopyBuf::mock).take(42);

        zero_buf_op(none_buf.into()).unwrap();
        maybe_buf_op(some_buf.into()).unwrap();
        one_buf_op(one_buf.into()).unwrap();
        multi_buf_op(multi_bufs.into());
    }

    #[test]
    fn test_buf_list_err() {
        let none_buf = Option::<ZeroCopyBuf>::None;
        let some_buf = Some(ZeroCopyBuf::mock());
        let many_bufs = repeat_with(ZeroCopyBuf::mock).take(42);

        zero_buf_op(some_buf.into()).unwrap_err();
        maybe_buf_op(many_bufs.into()).unwrap_err();
        one_buf_op(none_buf.into()).unwrap_err();
    }
}
```

**Comment by SyrupThinker** [comment]
This makes the interface a lot more complicated. `ZeroCopyBuf` is backed by `SharedRef`, couldn't `ZeroCopyBuf` implement `Clone` (internally incrementing the reference count) so the user just `clone`'s it to extend its lifetime?

**Comment by piscisaureus** [comment]
> This makes the interface a lot more complicated. `ZeroCopyBuf` is backed by `SharedRef`, couldn't `ZeroCopyBuf` implement `Clone` (internally incrementing the reference count) so the user just `clone`'s it to extend its lifetime?

Yes, that's definitely possible. I think al it needs is `#[derive(Clone)]`.
Cloning a SharedRef is not entirely free but I think it's cheap enough.

File: core/bindings.rs [hunk 1]
```diff
@@ -452,17 +452,46 @@ fn send(
...
+              let msg = v8::String::new(scope, &s).unwrap();
+              scope.isolate().throw_exception(msg.into());
+              return;
+            }
+          },
+        );
+      }
+      state.dispatch_op(scope, op_id, control, &mut zero_copy_bufs)
+    }
+  };
```
Comment (SyrupThinker) [inline core/bindings.rs:494]: Any advice on how to reasonably reduce duplication here?

**Comment by SyrupThinker** [comment]
<details>
<summary>Benchmarks</summary>

## deno_tcp
### pre
```
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   188.01us   56.73us   3.15ms   87.62%
    Req/Sec    24.00k     1.37k   26.39k    80.35%
Requests/sec:  47731.27
Transfer/sec:      2.32MB
```
### post
```
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   185.99us   56.49us   3.25ms   87.55%
    Req/Sec    24.20k     1.40k   33.10k    88.53%
Requests/sec:  48007.72
Transfer/sec:      2.33MB
```
## http_benchmark testing DENO using net/http.
### pre
```
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   484.91us  297.98us  20.13ms   94.92%
    Req/Sec     9.87k     1.66k   16.52k    89.53%
Requests/sec:  19592.97
Transfer/sec:      2.07MB
```
### post
```
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   462.13us  215.14us  15.78ms   96.05%
    Req/Sec    10.26k     1.16k   11.53k    80.85%
Requests/sec:  20405.30
Transfer/sec:      2.16MB
```
## http_proxy_benchmark testing DENO using net/tcp.
### pre
```
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   526.90us  117.95us   6.27ms   90.48%
    Req/Sec     9.19k   510.24    10.32k    76.37%
Requests/sec:  18281.36
Transfer/sec:      1.53MB
```
### post
```
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   540.92us  178.53us  12.01ms   96.98%
    Req/Sec     9.00k   592.57    10.36k    83.25%
Requests/sec:  17913.28
Transfer/sec:      1.50MB
```
## http_benchmark testing deno_core_http_bench
### pre
```
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   161.79us   49.14us   2.72ms   85.19%
    Req/Sec    27.45k     1.63k   30.58k    78.50%
Requests/sec:  54594.64
Transfer/sec:      2.66MB
```
### post
```
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   163.57us   48.04us   2.41ms   84.09%
    Req/Sec    27.16k     1.62k   30.21k    82.75%
Requests/sec:  53998.81
Transfer/sec:      2.63MB
```
## http_benchmark testing RUST hyper.
### pre
```
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   136.87us   64.04us   4.24ms   79.92%
    Req/Sec    27.75k     1.80k   31.62k    68.41%
Requests/sec:  55169.35
Transfer/sec:      4.63MB
```
### post
```
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   135.82us   66.01us   4.90ms   82.32%
    Req/Sec    27.89k     1.87k   31.29k    70.00%
Requests/sec:  55464.29
Transfer/sec:      4.65MB
```

</details>

File: core/bindings.rs (hunk 1 above)
Comment (piscisaureus) [reply core/bindings.rs:494]: Not so obvious indeed. This is what I could think of:

```rust
  let mut buf_iter = (2..args.length()).map(|i| {
    v8::Local::<v8::ArrayBufferView>::try_from(args.get(i))
      .map(ZeroCopyBuf::new)
      .map_err(|err| {
        let msg = format!("Invalid argument at position {}: {}", i, err);
        v8::String::new(scope, &msg).unwrap()
      })
  });
  let mut buf_one: ZeroCopyBuf; // Initialized only when there is one buffer.
  let mut buf_vec: Vec<ZeroCopyBuf>; // Initialized only when there is more than one buffer.
  let buf_iter_result = match buf_iter.len() {
    0 => Ok(&mut [][..]),
    1 => match buf_iter.next().unwrap() {
      Ok(b) => {
        buf_one = b;
        Ok(std::slice::from_mut(&mut buf_one))
      }
      Err(e) => Err(e),
    },
    _ => match buf_iter.collect::<Result<Vec<_>, _>>() {
      Ok(v) => {
        buf_vec = v;
        Ok(&mut buf_vec[..])
      }
      Err(e) => Err(e),
    },
  };
  let maybe_response = match buf_iter_result {
    Ok(mut bufs) => state.dispatch_op(scope, op_id, control, &mut bufs),
    Err(msg) => {
      scope.isolate().throw_exception(msg.into());
      return;
    }
  };
```

Comment (piscisaureus) [reply core/bindings.rs:494]: I do think that it is a bit more efficient than pushing to a vector in a loop. However I'd be surprised if multi-buffer ops turn out to be a performance critical path.

**Comment by SyrupThinker** [comment]
> `.collect::<Result<Vec<_>, _>>()` should automatically do what your error filtering logic does.
> Just curious - was it not working for you?
> Too obscure?

Huh, I had dismissed that as a typo (: . Its noted in the last example of `collect`, that is quite obscure.

File: cli/ops/worker_host.rs [hunk 2]
```diff
@@ -345,11 +345,14 @@ fn op_host_get_message(
 fn op_host_post_message(
   state: &State,
   args: Value,
-  data: Option<ZeroCopyBuf>,
+  data: &mut [ZeroCopyBuf],
 ) -> Result<JsonOp, OpError> {
+  if data.len() != 1 {
+    panic!("Invalid number of arguments");
+  }
```
Comment (ry) [inline cli/ops/worker_host.rs:352]: Use assert!

**Comment by SyrupThinker** [comment]
> In general this looks good and I'm in favor of landing. However I don't see any tests which exercise multiple zero copy buffers being used... or have I missed it?

No you didn't miss it. Where would I add this test?
I have changed the one in test_plugin, is there another place where I should add one?

**Comment by hayd** [comment]
Is a good example http server / body in a single op?

**Comment by SyrupThinker** [comment]
> Is a good example http server / body in a single op?

I'm not sure what you mean

File: core/core_isolate.rs [hunk 3]
```diff
@@ -816,6 +829,48 @@ pub mod tests {
...
+  #[test]
+  fn test_dispatch_two_zero_copy_bufs() {
+    let (mut isolate, dispatch_count) = setup(Mode::AsyncZeroCopy(2));
+    js_check(isolate.execute(
+      "filename.js",
+      r#"
+        let control = new Uint8Array([24]);
+        let zero_copy_a = new Uint8Array([0]);
+        let zero_copy_b = new Uint8Array([1]);
+        Deno.core.send(1, control, zero_copy_a, zero_copy_b);
```
Comment (ry) [inline core/core_isolate.rs:868]: Nice work!

**Pull request merged**